### PR TITLE
Validate alias names the same as index names

### LINKED ISSUE #20748: Special characters allowed for aliases which are not allowed in index names
**Elasticsearch version**: 5.0.0-alpha5

**Plugins installed**: [] x-pack

**JVM version**:1.8.0_101

**OS version**: Ubuntu 14.04

**Description of the problem including expected versus actual behavior**:

**Steps to reproduce**:

It is possible to use special characters, e.g. `*` in aliases which are not allowed in Index names.  I think this is a bug as it can lead to very confusing states - below is an example:

```
PUT /foo
PUT /bar

# This will fail with invalid_index_name_exception
PUT /foo*

PUT /foo/bar/1
{
  "message": "Indexed into /foo/bar/1"
}

PUT /bar/bar/1
{
  "message":"Idexed into /bar/bar/1"
}

# This will fail with invalid_index_name_exception
PUT /bar*/bar/2
{
  "message":"Indexed into /bar*/bar/2"
}

POST /_aliases
{
  "actions":[
    {"add": { "index": "foo", "alias":"bar*" } }
    ]
}

# Now this will be successful - confusing!
PUT /bar*/bar/2
{
  "message":"Indexed into /bar*/bar/2"
}

# A search on indexes /bar* responds with results both from the index `bar` and from `foo` as a result of the alias.
GET /bar*/_search
{
  "took": 3,
  "timed_out": false,
  "_shards": {
    "total": 10,
    "successful": 10,
    "failed": 0
  },
  "hits": {
    "total": 3,
    "max_score": 1,
    "hits": [
      {
        "_index": "foo",
        "_type": "bar",
        "_id": "2",
        "_score": 1,
        "_source": {
          "message": "Indexed into /bar*/bar/2"
        }
      },
      {
        "_index": "bar",
        "_type": "bar",
        "_id": "1",
        "_score": 1,
        "_source": {
          "message": "Idexed into /bar/bar/1"
        }
      },
      {
        "_index": "foo",
        "_type": "bar",
        "_id": "1",
        "_score": 1,
        "_source": {
          "message": "Indexed into /foo/bar/1"
        }
      }
    ]
  }
}
```

### PR DESCRIPTION
**Pull request by nik9000**
Applied (almost) the same rules we use to validate index names
to new alias names. The only rule not applies it "must be lowercase".
We have tests that don't follow that rule and I except there are lots
of examples of camelCase alias names in the wild. We can add that
validation but I'm not sure it is worth it.

Closes #20748

### CODE REVIEW

**Comment by nik9000** [comment]
I'm unsure if this is a thing that we should backport to 5.x because it is breaking. It might have to come with 6.0.

I'll add a breaking changes not to some document but I'd like to wait to see if we want this in 5.x first.

**Comment by s1monw** [comment]
this LGTM but I wonder what happens if somebody changes an existing alias that doesn't pass this check? I also think we should add it to 5.1 to be honest

**Comment by nik9000** [comment]
Yeah, I think I ought to add some more tests around backwards compatibility. Changing an existing alias and just working with aliases that don't pass the test in general.

I'm fairly sure any attempt to change an alias that doesn't pass the test will just fail, but it is worth poking at further.

**Comment by nik9000** [comment]
@s1monw do you have an opinion on the lowercase check? I didn't add it because I thought it'd invalidate a ton of aliases folks use and it doesn't add the same value as preventing stuff like `-` and `*` does.

**Comment by s1monw** [comment]
I am good with skipping the lowercase part

**Comment by nik9000** [comment]
Update: Existing aliases work fine. You can start the cluster with aliases that you aren't allowed to create without an issue. I expect I should add a now-invalid alias to the backwards compatibility index creation scripts so we have an assertion about that.

Also modifying an alias isn't really a thing so far as I can tell. I mean, I don't know *all* of this code but I know a fair bit of it and I think the way to modify an alias is to remove and re-add it. So if we merged this then you couldn't "modify" an alias that failed the validation.

**Comment by nik9000** [comment]
@s1monw I added some backwards compatibility tests: you can read from "bad" aliases and remove them but you can't add any more. That means you can't edit them.

**Comment by nik9000** [comment]
Question: should I add this to the breaking changes docs in 6.0 *and* in 5.1 or just in 5.1? I don't really like breaking 5.1 at all but this is fairly "bug fix"-ish.

**Comment by javanna** [comment]
I would consider this as a (bad) bug. For instance, up until now it was possible to create an alias that contains `*` which messes up wildcard filtering in get aliases api.

**Comment by nik9000** [comment]
It is fairly nasty, yeah. An alias with * in the name gets you very weird
results. Which is why 5.1 seems appropriate even though this is technically
a breaking change.

**Comment by nik9000** [comment]
Anyone want to review the BWC tests?

**Comment by nik9000** [comment]
@javanna can you have a look at the BWC testing on this? I already got a positive code review on the PR but I since wrote a bunch of BWC tests for it.

**Review by javanna — APPROVED** [review]
left some minor comments, LGTM otherwsie

File: core/src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityIT.java [hunk 1]
```diff
@@ -428,6 +430,24 @@ void assertPositionIncrementGapDefaults(String indexName, Version version) throw
...
+     * search-able though.
+     */
+    void assertAliasWithBadName(String indexName, Version version) throws Exception {
+        if (version.onOrAfter(Version.fromString("5.1.0-alpha1"))) {
+            return;
+        }
+        // We can read from the alias just like we can read from the index.
+        String aliasName = "#" + indexName;
+        long totalDocs = client().prepareSearch(indexName).setSize(0).get().getHits().totalHits();
+        assertHitCount(client().prepareSearch(aliasName).setSize(0).get(), totalDocs);
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityIT.java]: It asserts the hit count now. Is that enough?

Comment (javanna) [inline core/src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityIT.java]: Did you mean to push some commit? I don't see any change here

File: core/src/test/java/org/elasticsearch/cluster/metadata/AliasValidatorTests.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        assertEquals("Invalid alias name [_cat]: must not start with '_', '-', or '+'", e.getMessage());
+        e = expectThrows(InvalidAliasNameException.class, () -> validator.validateAliasStandalone("-cat", null));
+        assertEquals("Invalid alias name [-cat]: must not start with '_', '-', or '+'", e.getMessage());
+        e = expectThrows(InvalidAliasNameException.class, () -> validator.validateAliasStandalone("+cat", null));
+        assertEquals("Invalid alias name [+cat]: must not start with '_', '-', or '+'", e.getMessage());
+        e = expectThrows(InvalidAliasNameException.class, () -> validator.validateAliasStandalone("c*t", null));
+        assertThat(e.getMessage(), startsWith("Invalid alias name [c*t]: must not contain the following characters "));
+
+        // Doesn't throw an exception
+        validator.validateAliasStandalone("CAT", null);
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/cluster/metadata/AliasValidatorTests.java]: I add a comment and can open a followup discuss issue, sure!

File: core/src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityIT.java (hunk 1 above)
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityIT.java]: It already asserted it.

**Comment by s1monw** [comment]
@nik9000 do you wanna push this?

File: core/src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityIT.java (hunk 1 above)
Comment (javanna) [inline core/src/test/java/org/elasticsearch/bwcompat/OldIndexBackwardsCompatibilityIT.java]: you are asserting that the two hit counts returned are the same. Couldn't they be both `0` in theory?

**Comment by nik9000** [comment]
Sorry this took so long to push!

master: 3787ea27bf1b33b8005547969b358b8d19fddc8f
5.x: in progress

**Comment by nik9000** [comment]
All merged! Thanks to @javanna and @s1monw for being so patient with this one.

**Pull request merged**